File: Daemon/SCDaemonXPC.m
//
//  SCDaemonXPC.m
//  selfcontrold
//
//  Created by Charlie Stigler on 5/30/20.
//

#import "SCDaemonXPC.h"
#import "SCDaemonBlockMethods.h"
#import "SCXPCAuthorization.h"

@implementation SCDaemonXPC

- (void)startBlockWithControllingUID:(uid_t)controllingUID blocklist:(NSArray<NSString*>*)blocklist isAllowlist:(BOOL)isAllowlist endDate:(NSDate*)endDate blockSettings:(NSDictionary*)blockSettings authorization:(NSData *)authData reply:(void(^)(NSError* error))reply {
    NSLog(@"XPC method called: startBlockWithControllingUID");
    
    NSError* error = [SCXPCAuthorization checkAuthorization: authData command: _cmd];
    if (error != nil) {
        if (![SCMiscUtilities errorIsAuthCanceled: error]) {
            NSLog(@"ERROR: XPC authorization failed due to error %@", error);
            [SCSentry captureError: error];
        }
        reply(error);
        return;
    } else {
        NSLog(@"AUTHORIZATION ACCEPTED for startBlock with authData %@ and command %s", authData, sel_getName(_cmd));
    }

    [SCDaemonBlockMethods startBlockWithControllingUID: controllingUID blocklist: blocklist isAllowlist:isAllowlist endDate: endDate blockSettings:blockSettings authorization: authData reply: reply];
}

- (void)updateBlocklist:(NSArray<NSString*>*)newBlocklist authorization:(NSData *)authData reply:(void(^)(NSError* error))reply {
    NSLog(@"XPC method called: updateBlocklist");
    
    NSError* error = [SCXPCAuthorization checkAuthorization: authData command: _cmd];
    if (error != nil) {
        if (![SCMiscUtilities errorIsAuthCanceled: error]) {
            NSLog(@"ERROR: XPC authorization failed due to error %@", error);
            [SCSentry captureError: error];
        }
        reply(error);
        return;
    } else {
        NSLog(@"AUTHORIZATION ACCEPTED for updateBlocklist with authData %@ and command %s", authData, sel_getName(_cmd));
    }
    
    [SCDaemonBlockMethods updateBlocklist: newBlocklist authorization: authData reply: reply];
}

- (void)updateBlockEndDate:(NSDate*)newEndDate authorization:(NSData *)authData reply:(void(^)(NSError* error))reply {
    NSLog(@"XPC method called: updateBlockEndDate");
    
    NSError* error = [SCXPCAuthorization checkAuthorization: authData command: _cmd];
    if (error != nil) {
        if (![SCMiscUtilities errorIsAuthCanceled: error]) {
            NSLog(@"ERROR: XPC authorization failed due to error %@", error);
            [SCSentry captureError: error];
        }
        reply(error);
        return;
    } else {
        NSLog(@"AUTHORIZATION ACCEPTED for updateBlockENdDate with authData %@ and command %s", authData, sel_getName(_cmd));
    }
    
    [SCDaemonBlockMethods updateBlockEndDate: newEndDate authorization: authData reply: reply];
}

// Part of the HelperToolProtocol.  Returns the version number of the tool.  Note that never
// requires authorization.
- (void)getVersionWithReply:(void(^)(NSString * version))reply {
    NSLog(@"XPC method called: getVersionWithReply");
    // We specifically don't check for authorization here.  Everyone is always allowed to get
    // the version of the helper tool.
    reply(SELFCONTROL_VERSION_STRING);
}

@end

File: Daemon/SCDaemonXPC.swift
import Foundation

/// Handles incoming XPC calls. Each connection gets its own instance.
/// Validates authorization, then delegates to SCDaemonBlockMethods.
final class SCDaemonXPC: NSObject, SCDaemonProtocol {

    func startBlock(controllingUID: UInt32,
                    blocklist: [String],
                    isAllowlist: Bool,
                    endDate: Date,
                    blockSettings: [String: Any],
                    authorization: Data,
                    reply: @escaping (Error?) -> Void) {
        do {
            try SCXPCAuthorization.checkAuthorization(authorization, for: "startBlock")
            try SCDaemonBlockMethods.shared.startBlock(
                controllingUID: uid_t(controllingUID),
                blocklist: blocklist,
                isAllowlist: isAllowlist,
                endDate: endDate,
                blockSettings: blockSettings
            )
            reply(nil)
        } catch {
            NSLog("stonectld: startBlock failed: %@", error.localizedDescription)
            reply(error)
        }
    }

    func updateBlocklist(_ newBlocklist: [String],
                         authorization: Data,
                         reply: @escaping (Error?) -> Void) {
        do {
            try SCXPCAuthorization.checkAuthorization(authorization, for: "updateBlocklist")
            try SCDaemonBlockMethods.shared.updateBlocklist(newBlocklist)
            reply(nil)
        } catch {
            reply(error)
        }
    }

    func updateBlockEndDate(_ newEndDate: Date,
                            authorization: Data,
                            reply: @escaping (Error?) -> Void) {
        do {
            try SCXPCAuthorization.checkAuthorization(authorization, for: "updateBlockEndDate")
            try SCDaemonBlockMethods.shared.updateBlockEndDate(newEndDate)
            reply(nil)
        } catch {
            reply(error)
        }
    }

    func getVersion(reply: @escaping (String) -> Void) {
        reply(StoneConstants.versionString)
    }
}
